desktop_app/src/ui/contexts/multi-chat-manager.tsx

```diff
@@ -0,0 +1,385 @@
+import { useChat } from '@ai-sdk/react';
+import { DefaultChatTransport, UIMessage } from 'ai';
+import { createContext, useCallback, useContext, useEffect, useMemo, useRef, useState } from 'react';
+
+import config from '@ui/config';
+import { useMessageActions } from '@ui/hooks/use-message-actions';
+import { getAllMemories } from '@ui/lib/clients/archestra/api/gen';
+import { useChatStore, useCloudProvidersStore, useOllamaStore, useToolsStore } from '@ui/stores';
+import { useStatusBarStore } from '@ui/stores/status-bar-store';
+
+interface ChatInstanceState {
+  sessionId: string;
+  chatId: number;
+  title: string;
+  messages: UIMessage[];
+  status: string;
+  isLoading: boolean;
+  isSubmitting: boolean;
+  editingMessageId: string | null;
+  editingContent: string;
+  regeneratingIndex: number | null;
+  fullMessagesBackup: UIMessage[];
+  hasTooManyTools: boolean;
+  hasLoadedMemories: boolean;
+}
+
+interface ChatInstanceActions {
+  setMessages: (msgs: UIMessage[]) => void;
+  sendMessage: (args: { text: string }) => void;
+  stop: () => void;
+  regenerate: () => void;
+  setIsSubmitting: (b: boolean) => void;
+  setEditingContent: (c: string) => void;
+  startEdit: (id: string, content: string) => void;
+  cancelEdit: () => void;
+  saveEdit: (id: string) => Promise<void>;
+  deleteMessage: (id: string) => Promise<void>;
+  handleRegenerateMessage: (idx: number) => Promise<void>;
+  setHasTooManyTools: (b: boolean) => void;
+  setHasLoadedMemories: (b: boolean) => void;
+  loadMemoriesIfNeeded: () => Promise<void>;
+}
+
+export type ChatInstance = ChatInstanceState & ChatInstanceActions;
+
+interface MultiChatManagerContextType {
+  getCurrentChatInstance: () => ChatInstance | null;
+  getChatInstance: (sessionId: string) => ChatInstance | null;
+}
+
+const MultiChatManagerContext = createContext<MultiChatManagerContextType | null>(null);
+
+// Component that manages a single chat instance
+function ChatInstanceManager({
+  sessionId,
+  chatId,
+  title,
+  onInstanceCreated,
+  initialMessages = [],
+}: {
+  sessionId: string;
+  chatId: number;
+  title: string;
+  onInstanceCreated: (instance: ChatInstance) => void;
+  initialMessages?: UIMessage[];
+}) {
+  const { getCurrentChat } = useChatStore();
+  const { selectedToolIds } = useToolsStore();
+  const { selectedModel } = useOllamaStore();
+  const { availableCloudProviderModels } = useCloudProvidersStore();
+  const { setChatInference } = useStatusBarStore();
+  const [hasTooManyTools, setHasTooManyTools] = useState(false);
+  const [hasLoadedMemories, setHasLoadedMemories] = useState(false);
```

Comment:
could we use the `useMemoryStore` instead of having to worry about fetching/refetching memories here?

I believe memories are fetched on app load and stored in a `zustand` store, so we _shouldn't_ need to refetch them